design_documents/parametrizing_tests_with_fixed_values.md

```diff
@@ -0,0 +1,143 @@
+
+## Proposed Solution
+
+Introduce new attribute `test_case`.
+
+Generate code for each of the test cases so no handling of values injection, etc. is necessary.
+The compiler would then validate if the values for the generated code are actually valid.
+Thanks to that, we could have failures for invalid arguments early in the execution.
+
+### Parametrization With Arguments
+
+```cairo
+#[test_case(1, 3)]
+#[test_case(3, 5)]
+fn my_test(a: felt252, b: u32) {
+    // ...
+}
+```
+
+For test cases like this, we would generate two actual test cases:
+
+```cairo
+#[test]
+fn my_test_case_1() {
+  let a: felt252 = 1;
+  let b: u32 = 3;
+  // ...
+}
+
+#[test]
+fn my_test_case_2() {
+    let a: felt252 = 3;
+    let b: u32 = 5;
+  // ...
+}
+```
+
+### Generated Cases Names
+
+Generated test cases should be named in such a way that we can automatically detect which tests are generated cases of
+the same base test.
+
+Test in `snforge` can be filtered by name.
+It is important that generated test names do not break the test filtering logic.
+
+This can be resolved by either using names for generated tests that still work with filters as the base test would do,
+or changing the filtering logic, so it can recognize generated test cases and treat them accordingly.
+
+### Parametrizing With Complex Types
+
+Complex types would result in the same code generation as simple types:
+
+```cairo
+#[test_case(MyStruct { a: 1, b: array![2, 3] })]
+fn my_test(a: MyStruct) {
+    // ...
+}
+```
+
+```cairo
+#[test]
+fn my_test_1() {
+    let a = MyStruct { a: 1, b: array![2, 3] };
+    // ...
+}
+```
+
+### Problems With Code Generation
+
+The code generation solution has some problems that will need to be addressed before it is implemented.
+See [section below](#possible-problems-with-code-generation) for details.
+
+## User Experience
+
+Unlike fuzz-tests, we should run and indicate the result of all parametrized cases.
+If some test cases fail, others should still be executed and their results should be displayed to the user.
+
+An example output could look similarly to this:
+
+```shell
+$ snforge test
+[PASS] tests::parametrized(a = 1, b = 2, c = my_fixture(1, 2))
+[PASS] tests::parametrized(a = 3, b = 5, c = my_fixture(4, 5))
+[FAIL] tests::parametrized(a = 4, b = 5, c = my_fixture(3, 2))
+
+Failure data:
+    original value: [344693033283], converted to a string: [PANIC]
+    
+[PASS] tests::parametrized(a = 5, b = 7, c = my_fixture(7, 8))
+# ...
+```
+
+## Deterministic Test Execution Order
+
+Parallel test execution implementation in forge makes deterministic test execution order non-trivial.
+For good user experience, we should aim to implement parametrized tests in such a way that cases are executed in
+deterministic order directly after each other.
+
+This is not a technical requirement for implementing parametrized tests but will make for a much better UX, especially
+in case of displaying failures of some parametrized cases.
```

Comment:
I think the problem here is not in the order but in output collection.
Ideally, we could be able to execute cases in parallel, collect the output and print it after all cases are finished

Reply:
Correct, they could probably run in any order as long as we output them in deterministic manner. Handling of this would not be obvious in case `--exit-first` flag is used however.

Reply:
Probably, but we should investigate it because we want those tests to be fast

Reply:
I'll update this section so it explains we only want results displayed in deterministic order.

Reply:
I think if we enforce naming the tests in the attribute (as you described above) we can just treat each case as a separate test and everything will be fine

Reply:
I don't want to enforce naming things, this is inflexible. And for me parametric tests only make sense if there are treated as a single "bundle" in output.

Reply:
`test_case` crate treats them as a separate entities, example output when testing https://github.com/foundry-rs/starknet-foundry/blob/93b2a2a976073411d3ec694dcc330b67f1b8ab3b/crates/cheatnet/tests/starknet/cheat_fork.rs#L135-L137
```console
running 2 tests
test starknet::cheat_fork::warp_cairo0_contract::when_common_call ... ok
test starknet::cheat_fork::warp_cairo0_contract::when_library_call ... ok
```
I think we could stick to that, I don't see any major UX benefit of the [pytest-like approach](https://docs.pytest.org/en/7.1.x/how-to/parametrize.html#pytest-mark-parametrize-parametrizing-test-functions), I guess it's a matter of taste

Reply:
It's a matter of preference, but for me only argument for not bundling tests together is easier implementation. From the UX side there are no benefits.

Reply:
I agree, but if it is a matter of preference then easier implementation becomes an argument I guess (even though it really shouldn't, but since there are not other arguments...)

Reply:
From code generation standpoint, both versions are pretty much te same. It's just a case of collecting and displaying them together in some logical manner.